File: algorithm/42. Trapping Rain Water/42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int res=0, l=0, r=height.size()-1;
        int minh;
        while(l<r)
        {
            minh=min(height[l],height[r]);
            if(minh==height[l])
            {
                while(++l<r && height[l]<minh)
                {
                    res+=minh-height[l];
                }
            }
            else
            {
                while(--r>l && height[r]<minh)
                {
                    res+=minh-height[r];
                }
            }
        }
        return res;
    }
};

File: algorithm/42. Trapping Rain Water/42.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        minh,l,r,res=0,0,len(height)-1,0
        while l<r:
            minh=min(height[l],height[r])
            if height[l]==minh:
                l+=1
                while l<r and height[l] < minh:
                    res+=minh-height[l]
                    l+=1
            else:
                r-=1
                while l<r and height[r]<minh:
                    res+=minh-height[r]
                    r-=1
        return res
